src/main/java/io/vertx/mqtt/impl/MqttEndpointImpl.java

```diff
@@ -438,11 +440,19 @@ public MqttEndpointImpl publish(String topic, Buffer payload, MqttQoS qosLevel,
 
   @Override
   public MqttEndpointImpl publish(String topic, Buffer payload, MqttQoS qosLevel, boolean isDup, boolean isRetain, Handler<AsyncResult<Integer>> publishSentHandler) {
+    return publish(topic, payload, qosLevel, isDup, isRetain, this.nextMessageId(), publishSentHandler);
+  }
+
+  @Override
+  public MqttEndpointImpl publish(String topic, Buffer payload, MqttQoS qosLevel, boolean isDup, boolean isRetain, int messageId, Handler<AsyncResult<Integer>> publishSentHandler) {
+    if(messageId > this.messageIdCounter % MAX_MESSAGE_ID || messageId < 0) {
```

Comment:
let's think about the case when the MQTT server crash but for handling sessions restore you have stored the latest messageId somewhere. On restarting the `this.messageIdCounte` will be zero so the first condition will not check that you are not setting a value not higher then MAX_MESSAGE_ID because the module will be always 0, right? The `messageIdCounter` is used only if you use the `nextMessageId` so we should not use it here or we should have a way for setting it and then after recovering you can continue to use the publish leveraging the `nextMessageId`. Wdyt? What's the case you are working on?

Reply:
Sorry, there was a dumb mistake in the range validation code. Please take a look at the updated version.

Reply:
Yeah now it's better but I would like to know how you are going to use this feature ... I mean, you now have the possibility to set the `messageId` so are you going to use always this way or do you think that after recovering a session (from a storage) so sending all the pending messages ids, is useful to set the `messageIdCounter` and relying on the automatic `nextMessageId`? Just thinking aloud ... depending on the use case you have.

Reply:
In my use case, messageId will always be specified manually, from a persistent external source. If you think that it might be useful - I can also change it to update `messageIdCounter` whenever `messageId` is provided manually.

Reply:
Well, I don't think that the counter should be update on publish because publishing messages could even happens not in the order of the message id. It could happen in some cases that messages are published with different QoS (1 or 2, where messageId makes sense) and someone want to just to procede to the next message if sending the current one fails somehow. It means that on retry, the message counter is set to an old previous message id in this case. Let's think more about that but for a different PR if needed.